confgenerator/nginx.go

```diff
@@ -0,0 +1,51 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package confgenerator
+
+import "github.com/GoogleCloudPlatform/ops-agent/confgenerator/otel"
+
+type MetricsReceiverNginx struct {
+	ConfigComponent `yaml:",inline"`
+
+	MetricsReceiverShared `yaml:",inline"`
+
+	Endpoint string `yaml:"endpoint" validate:"required,url"`
```

Comment:
@jsuereth As I mentioned to you I don't love the name `endpoint` because it doesn't help the user figure out what it should be set to. (e.g. something like `status_url` would be clearer)

@djaglowski Is there not a sane default for this? `http://localhost/status` perhaps?

Reply:
Regarding the name endpoint:
1. `endpoint` is what was chosen by OTel prior to this project.  
2. [Prometheus exporter](https://github.com/nginxinc/nginx-prometheus-exporter#running-the-exporter-binary) uses a vey non-ops-agent, non-otel looking endpoint.   They do recommend the user use a default exposed path of "stub_status".
3. `endpoint` is canonical in most receivers in otel.

If you think Ops Agent should deviate, that's worth a discussion.  I do not think we should look significantly different from otel without a good reason.



Regarding you question on sane default, the answer is not really.  users *have* to configure the plugin.  The [default documentation](https://nginx.org/en/docs/http/ngx_http_stub_status_module.html#stub_status) recommends (`/status`) which likely conflicts with a lot of hosted web applications.  

We have two options:

1. We can default to `/status` because this is what the default set up documentation suggests.
2. We can default to `/sub_status` because this is the default recommended by prometheus.
3. We can not have a default since a user *has* to configure this URL.

I'd vote for option #1 (preffered) with #2 as my second.

Reply:
I agree we should stick with `endpoint`. This will be the closest thing to a "universal" field across 100+ receivers, and I think that providing users with a familiar anchor point for each config is more useful than hinting at detailed meaning. 

Also would second that #1 is the best default because it aligns with the set up docs.

Reply:
I don't see the value in a "universal" field - especially one that might contain an HTTP URL for one receiver, a Unix socket path for another, and a database connection string for a third. Unless you're saying `endpoint` always means a URL? I'm happy to punt this discussion to after we have the first working demo.

Reply:
endpoint always means a URL in opentelemetry collector.  I think the discussion on URL vs. Endpoint should be amongst @qingling128 and team, outside the scope of this PR.   Can you schedule/make sure that happens Quentin?

Also, found another possible default for status, the current cloud docs recommend [this config](https://raw.githubusercontent.com/Stackdriver/stackdriver-agent-service-configs/master/etc/nginx/conf.d/status.conf) for nginx, which means we probably should default to `/nginx_status` to ease migration.   We do *not* have a discovery mechanism to *look* for 443/80 port openings.

Given all that, (1) or (2) as default seems fine for now. I think we can (in a non-breaking way) improve the default overtime as long as we don't regress the behavior.

Reply:
@quentinmit, can you double-check / correct me on this?

---

I've removed the `required` tag and added the default as follows:
```go
type MetricsReceiverNginx struct {
	ConfigComponent `yaml:",inline"`
	MetricsReceiverShared `yaml:",inline"`
	Endpoint string `yaml:"endpoint" validate:"url=http://localhost/status"`
}
```

This should invalidate several of the golden error cases that I added (they should not generate any error), but I'm getting the following error instead: `the agent config file is not valid. detailed error: [19:16] Key: 'MetricsReceiverNginx.endpoint' Error:Field validation for 'endpoint' failed on the 'url' tag`

I believe the validator is disregarding the absence of the `required` tag, and enforcing that `""` must be a valid `url`, which of course it is not.

---

The same behavior can be replicated on `master` with `collection_interval`, by removing the `required` tag from `MetricsReceiverShared.CollectionInterval`, and correspondingly the `collection_interval` from a test config, such as `testdata/valid/*/metrics-receiver_custom_collection_interval/input.yaml`:
```
> go test -mod=mod github.com/GoogleCloudPlatform/ops-agent/confgenerator
--- FAIL: TestGenerateConfsWithValidInput (0.00s)
    --- FAIL: TestGenerateConfsWithValidInput/linux (0.00s)
        --- FAIL: TestGenerateConfsWithValidInput/linux/metrics-receiver_custom_collection_interval (0.00s)
            confgenerator_test.go:134: MergeConfFiles("testdata/valid/linux/metrics-receiver_custom_collection_interval/built-in-config.yaml", "testdata/valid/linux/metrics-receiver_custom_collection_interval/input.yaml", "testdata/valid/linux/metrics-receiver_custom_collection_interval/merged-config.yaml") got: the agent config file is not valid. detailed error: Key: 'MetricsReceiverShared.collection_interval' Error:Field validation for 'collection_interval' failed on the 'duration' tag
FAIL
FAIL    github.com/GoogleCloudPlatform/ops-agent/confgenerator  0.216s
FAIL
```

---

If I'm correct in my assessment here, then we either need to fix this quickly, or accept that we'll not have a default in this first release.

Reply:
I believe you're looking for the `omitempty` tag:

```
	Endpoint string `yaml:"endpoint" validate:"omitempty,url"`
```

and then you should replace `""` with the default value at config generation time. I'm not sure if there's a way to include the default value itself directly in the struct tags, but I'm pretty sure it's not with `url=http://foo`.

See https://pkg.go.dev/github.com/go-playground/validator/v10#hdr-Omit_Empty

Reply:
Thanks Quentin. Now using `omitempty` in place of `required`, rather than just removing `required`. Default is set if necessary in the `Pipelines()` method.

Reply:
Summarizing the discussion offline.

**TL;DR** - We'll change the `endpoint` parameter name to `stub_status_url`.

**Longer story**
Nginx has various different plugins that users can use to expose metrics (today we support the `stub_status_module` only). e.g.
* http://nginx.org/en/docs/http/ngx_http_stub_status_module.html
* http://nginx.org/en/docs/http/ngx_http_api_module.html
* http://nginx.org/en/docs/http/ngx_http_status_module.html (commercial only)
* https://github.com/nginxinc/nginx-prometheus-exporter
* https://github.com/hnlq715/nginx-vts-exporter
* https://github.com/knyar/nginx-lua-prometheus

To ensure the config model is forward compatible, the options in the future in case we have to support some different flavors are:

Option 1 - Different Nginx related receiver for each of the plugins/modules above
Option 2 - Same Nginx receiver with different parameters that indicates the usage of different plugins/modules

We're leaning towards 2, while it should always be possible to do 1 if absolutely necessary. In order to reduce potential conflicts of what `endpoint` means for different Nginx modules, we'd like to make the parameter name specific to the metrics plugins/modules for forward compatibility.

Reply:
Follow up PR to address this is https://github.com/GoogleCloudPlatform/ops-agent/pull/166.